### DISCUSSION TITLE
Is there any one use prisma with another queryBuilder library like TypeORM in one project?

### DISCUSSION THREAD

**Discussion by SnowingFox** [prisma/prisma#23475] [Q&A]
### Question

Is there any one use prisma with another queryBuilder library like TypeORM in one project?

### How to reproduce (optional)

We know that in some cases, Prisma may not be as flexible as queryBuilder

for example, i have a relation like this, and i want get post likes count and is user liked current post.
```prisma
model User {
   likedPosts UserLikePosts[]
}

model Post {
    likes UserLikePosts[]
}


model UserLikePosts { ... }
```

we should query it like this
```ts
const data = prisma.post.find({
    where: { id: 1 },
    include: {
      _count: {
        select: {
          likes: true, 
        },
      },
      likes: {
        where: {
          userId: userId, 
        },
      },
    },
})
.then((data) => {
  type Data = typeof data & { isLiked: boolean }
  ;(data as Data).isLiked = Boolean(data.userLikePosts.length)
})
```

i think it's very ugly way to impl this example,

### Expected behavior (optional)

In typeorm, we can impl it like this
```ts
const data = await this.postRepo
      .createQueryBuilder('post')
      .setFindOptions({
        relations: {
          user: true,
        },
        where: {
          id,
        },
      })
      .loadRelationCountAndMap('hole.isLiked', 'post.likes', 'isLiked', (qb) =>
        qb.andWhere('isLiked.studentId = :studentId', {
          studentId: reqUser.studentId,
        }),
      )
      .getOne()
```

In some cases, Prisma is better than Typeorm, and vice versa, so is anybody to use typeorm (or queryBuilder lib) and prisma in one project？

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hi @SnowingFox 👋 

Thank you for raising this question.

Yes, it is possible to use Prisma Client in a project with another ORM like TypeORM. 

However, it's important to note that Prisma and TypeORM have different design philosophies and may not always work seamlessly together. For instance, TypeORM primarily leans on SQL operators for filtering lists or records, while Prisma provides a more generic set of operators that are intuitive to use. [docs](https://www.prisma.io/docs/orm/more/comparisons/prisma-and-typeorm#filtering).

If you're considering migrating from TypeORM to Prisma, Prisma provides a step-by-step guide on how to do [this](https://www.prisma.io/docs/orm/more/migrating-to-prisma/migrate-from-typeorm). This guide can also be useful if you're planning to use both ORMs in your project, as it can help you understand how to replace TypeORM queries with Prisma Client.

Remember, Prisma ORM lends itself really well for incremental adoption. This means, you don't have to migrate your entire project from TypeORM to Prisma ORM at once, but rather you can step-by-step move your database queries from [TypeORM to Prisma ORM](https://www.prisma.io/docs/orm/more/migrating-to-prisma/migrate-from-typeorm).

Do note that while it's technically possible to use Prisma and another ORM like TypeORM in the same project, it may require careful planning and consideration to ensure they work well together.

You can also take a look at [Prisma-Kysely](https://github.com/valtyr/prisma-kysely#readme). Kysely is another SQL Query builder that works seamlessly with Prisma with the Prisma-Kysely library. See this [blog post](https://bgolebiowski.com/blog/prisma-and-kysely-can-we-have-both).

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
